pkg/cdi/cdi.go

```diff
@@ -0,0 +1,105 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cdi
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	cdiSpecs "github.com/container-orchestrated-devices/container-device-interface/specs-go"
+	"github.com/golang/glog"
+
+	"github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/types"
+)
+
+// CreateCDISpec creates CDI spec file with specified devices
+func CreateCDISpec(resourcePrefix string, filteredDevices []types.HostDevice, rPool types.ResourcePool) error {
+	prefix := resourcePrefix
+	if prefixOverride := rPool.GetResourcePrefix(); prefixOverride != "" {
+		prefix = prefixOverride
+	}
+	cdiDevices := make([]cdiSpecs.Device, 0)
+	cdiSpec := cdiSpecs.Spec{
+		Version: cdiSpecs.CurrentVersion,
+		Kind:    prefix + "/net",
+		Devices: cdiDevices,
+	}
+
+	for _, dev := range filteredDevices {
+		containerEdit := cdiSpecs.ContainerEdits{
+			DeviceNodes: make([]*cdiSpecs.DeviceNode, 0),
+			Env:         make([]string, 0),
+		}
+
+		envs, err := rPool.GetEnvs(prefix, []string{dev.GetDeviceID()})
+		if err != nil {
+			glog.Errorf("CreateCDISpec(): Unable to create environment variables: %s", err)
+			return fmt.Errorf("unable to create environment variables: %s", err)
+		}
+		for name, value := range envs {
+			containerEdit.Env = append(containerEdit.Env, name+"="+value)
+		}
+
+		for i, spec := range dev.GetDeviceSpecs() {
+			glog.Infof("Checking device spec (%d): %s", i, spec.String())
+			deviceNode := cdiSpecs.DeviceNode{
+				Path:        spec.ContainerPath,
+				HostPath:    spec.HostPath,
+				Permissions: "rwm",
+			}
+			containerEdit.DeviceNodes = append(containerEdit.DeviceNodes, &deviceNode)
+		}
+		device := cdiSpecs.Device{
+			Name:           dev.GetDeviceID(),
+			ContainerEdits: containerEdit,
+		}
+		cdiSpec.Devices = append(cdiSpec.Devices, device)
+	}
+
+	err := cdi.GetRegistry().SpecDB().WriteSpec(&cdiSpec, prefix)
+	if err != nil {
+		glog.Errorf("CreateCDISpec(): can not create CDI json: %q", err)
+		return err
+	}
+
+	return nil
+}
+
+// CreateContainerAnnotations creates container annotations based on CDI spec for a container runtime
+func CreateContainerAnnotations(devicesIDs []string, cdiCache *cdi.Cache) (map[string]string, error) {
+	annotations := make(map[string]string, 0)
+	for _, id := range devicesIDs {
+		var cdiDevice *cdi.Device
+
+		for _, device := range cdiCache.ListDevices() {
```

Comment:
why cant you use `GetDevice()` directly ? what information are you missing ?

Reply:
GetDevice expects device ID in different format rather than device plugin provides

Reply:
The format is well defined and predictable, seeing you are the one who is writing this spec to begin with. so i think you can "reconstruct" it before calling getDevice.

if thats not the case, please explain and give example of the string format that GetDevice expects.

Reply:
refactored and fixed